qemu/tests/qemu_guest_agent.py

```diff
@@ -2043,6 +2020,95 @@ def result_check(flag, write_timeout, session):
         result_check("thaw", write_timeout, session)
 
+    @error_context.context_aware
+    def gagent_check_fstrim(self, test, params, env):
+        """
+        Execute "guest-fstrim" command via guest agent.
+
+        Steps:
+        1) boot up guest with scsi backend device
+        2) init and format the data disk
+        3) get the original blocks of data disk
+        4) create fragment in data disk
+        5) check the used blocks
+        6) execute fstrim cmd via guest agent
+        7) check if the used blocks is decreased
+
+        :param test: kvm test object
+        :param params: Dictionary with the test parameters
+        :param env: Dictionary with test environment.
+        """
+        def get_blocks():
+            """
+            Get the used blocks of data disk.
+            :return: the used blocks
+            """
+            blocks = process.system_output("stat -t %s" % stg_image_filename)
+            return blocks.strip().split()[2]
+
+        session = self._get_session(params, None)
+        self._open_session_list.append(session)
+
+        error_context.context("Format data disk.", logging.info)
+        image_size_disk = params["image_size_stg"]
+        disk_index = utils_misc.wait_for(
+            lambda: utils_disk.get_windows_disks_index(session,
+                                                       image_size_disk), 120)
+        if not disk_index:
+            test.error("Didn't get windows disk index.")
+        logging.info("Clear readonly of disk and online it in windows guest.")
+        if not utils_disk.update_windows_disk_attributes(session, disk_index):
+            test.error("Failed to update windows disk attributes.")
+        mnt_point = utils_disk.configure_empty_disk(
+            session, disk_index[0], image_size_disk, "windows",
+            labeltype="msdos")
+
+        error_context.context("Check the original blocks of data disk.",
+                              logging.info)
+        stg_image_params = params.object_params("stg")
+        stg_image_filename = storage.get_image_filename(
+            stg_image_params, data_dir.get_data_dir())
+        blocks_init = get_blocks()
+        logging.info("The blocks original is %s" % blocks_init)
+
+        error_context.context("Create fragment in data disk.", logging.info)
+        guest_dir = r"%s:" % mnt_point[0]
+        data_file = os.path.join(guest_dir,
+                                 "qga_fstrim%s" %
+                                 utils_misc.generate_random_string(5))
+        for i in range(5):
+            count = 1000 * (i + 1)
+            logging.info("Create %sM file in guest." % count)
+            cmd = "dd if=/dev/zero of=%s bs=1M count=%d" % (data_file, count)
+            session.cmd(cmd, timeout=600)
+            delete_file_cmd = "%s %s" % (params["delete_file_cmd"],
+                                         data_file.replace("/", "\\"))
```

Comment:
@xiagao  For windows,  This command( "del /f /q". ) is completely remove file from disk?  If it only put file to the recycling center.   Then it don't reduce disk size.

Reply:
I think it's removed from disk.And the following code verify it.